src/index.js

```diff
@@ -1,25 +1,55 @@
 'use strict';
 
-function wrap(schema, target) {
-  let proto = target.prototype;
-  let parent = Object.getPrototypeOf(target);
-  let staticProps = Object.getOwnPropertyNames(target);
-  let prototypeProps = Object.getOwnPropertyNames(proto);
-  let instanceProps = prototypeProps.filter(name => name !== 'constructor');
+function wrap(schema, target, hooks) {
+  hooks = hooks || [];
+  var proto = target.prototype;
+  var parent = Object.getPrototypeOf(target);
+  var staticProps = Object.getOwnPropertyNames(target);
+  var prototypeProps = Object.getOwnPropertyNames(proto);
+  var instanceProps = prototypeProps.filter(function (name) {
+    return name !== 'constructor';
+  });
+
+  // Wrap parent first
+  if (parent.name){
+    wrap(schema, parent, hooks);
+  }
+
+  // Add model schemas
+  if (target.schema && typeof target.schema == 'object') {
+    schema.add(target.schema);
+  }
+
+  // Only add a hook if it has not already been added
+  function addHook(type, hook){
+    let index = hooks.indexOf(hook);
+    if (index < 0){
+      hooks.push(hook)
+      hook(schema[type].bind(schema))
```

Comment:
The `hook` called here is a function defined in the class definition as bellow. I think my naming is confusing as in the example below `hook` is the parameter to this function.

```
Resource.hooks = {
  pre: [
    function(hook) {
      hook('save', function(next) {
        this.version += 1;
        next()
      })
    }
  ]
}
```